src/Agent.Plugins/BuildDrop/BuildDropDropPlugin.cs

```diff
@@ -0,0 +1,234 @@
+            ArgUtil.NotEmpty(projectId, nameof(projectId));
+
+            string buildIdStr = context.Variables.GetValueOrDefault(BuildVariables.BuildId)?.Value ?? string.Empty;
+            if (!int.TryParse(buildIdStr, out int buildId))
+            {
+                throw new ArgumentOutOfRangeException(buildIdStr);
+            }
+
+            string containerIdStr = context.Variables.GetValueOrDefault(BuildVariables.ContainerId)?.Value ?? string.Empty;
+            if (!long.TryParse(containerIdStr, out long containerId))
+            {
+                throw new ArgumentOutOfRangeException(buildIdStr);
+            }
+
+            string artifactName;
+            if (!context.Inputs.TryGetValue(ArtifactEventProperties.ArtifactName, out artifactName) ||
+                string.IsNullOrEmpty(artifactName))
+            {
+                throw new Exception(StringUtil.Loc("ArtifactNameRequired"));
+            }
+
+            string localPath;
+            if (!context.Inputs.TryGetValue(ArtifactEventProperties.TargetPath, out localPath) ||
+                string.IsNullOrEmpty(localPath))
+            {
+                throw new Exception(StringUtil.Loc("ArtifactLocationRequired"));
+            }
+
+            PreprocessResult result = CheckAndTransformTargetPath(context, localPath, artifactName);
+            if (result != null)
+            {
+                await ProcessCommandInternalAsync(context, projectId, buildId, artifactName, result, token);
+            }
+        }
+
+        // Run checks and conversions before processing. Returns null if the operation should be aborted.
+        protected abstract PreprocessResult CheckAndTransformTargetPath(
+            AgentTaskPluginExecutionContext context, string localPath, string artifactName);
+
+        // Process the command with preprocessed arguments.
+        protected abstract Task ProcessCommandInternalAsync(
+            AgentTaskPluginExecutionContext context, Guid projectId, int buildId, string artifactName, PreprocessResult result, CancellationToken token);
+    }
+
+    // To be invoked by PublishBuildArtifacts task
+    public class PublishBuildDropTask : BuildDropTaskPluginBase
+    {
+        // Same as: https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/PublishBuildArtifacts/task.json
+        public override Guid Id => new Guid("ECDC45F6-832D-4AD9-B52B-EE49E94659BE");
+
+        public override string Version => "0.137.0";
+
+        public override string Stage => "main";
+
+        protected override PreprocessResult CheckAndTransformTargetPath(
+            AgentTaskPluginExecutionContext context, string localPath, string artifactName)
+        { 
+            string hostType = context.Variables.GetValueOrDefault("system.hosttype")?.Value;
+            if (!IsUncSharePath(context, localPath) && !string.Equals(hostType, "Build", StringComparison.OrdinalIgnoreCase))
```

Comment:
why we check UNC here? why still restrict to "Build" the original reason is Release doesn't create FileContainer Id for each release.

Reply:
removed